File: apps/server/state/stream.py
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, field
from datetime import datetime


@dataclass
class StreamMessage:
    """Represents a stream chat message."""
    id: str
    content: str
    author: str
    timestamp: str = field(default_factory=lambda: datetime.utcnow().isoformat())
    superchat_amount: Optional[float] = None
    superchat_currency: Optional[str] = None
    is_member: bool = False


@dataclass
class StreamContext:
    """
    Manages the context of a live stream during workflow execution.
    """
    platform: Optional[str] = None  # "youtube", "twitch", "discord"
    video_id: Optional[str] = None
    channel_id: Optional[str] = None

    # Live stats
    viewer_count: int = 0
    like_count: int = 0

    # Queues
    message_queue: List[StreamMessage] = field(default_factory=list)
    superchat_queue: List[StreamMessage] = field(default_factory=list)

    # Timing
    stream_started_at: Optional[str] = None
    last_message_at: Optional[str] = None

    # Configuration
    max_queue_size: int = 100

    @property
    def silence_duration(self) -> float:
        """Get seconds since last message."""
        if not self.last_message_at:
            return 0
        last = datetime.fromisoformat(self.last_message_at)
        return (datetime.utcnow() - last).total_seconds()

    def add_message(self, message: StreamMessage):
        """Add a message to the queue."""
        self.message_queue.append(message)
        self.last_message_at = message.timestamp

        # Handle superchat
        if message.superchat_amount:
            self.superchat_queue.append(message)

        # Trim queues
        if len(self.message_queue) > self.max_queue_size:
            self.message_queue = self.message_queue[-self.max_queue_size:]
        if len(self.superchat_queue) > self.max_queue_size:
            self.superchat_queue = self.superchat_queue[-self.max_queue_size:]

    def pop_message(self) -> Optional[StreamMessage]:
        """Get and remove the next message from the queue."""
        if self.message_queue:
            return self.message_queue.pop(0)
        return None

    def pop_superchat(self) -> Optional[StreamMessage]:
        """Get and remove the next superchat from the queue."""
        if self.superchat_queue:
            return self.superchat_queue.pop(0)
        return None

    def update_stats(self, viewer_count: int = None, like_count: int = None):
        """Update stream statistics."""
        if viewer_count is not None:
            self.viewer_count = viewer_count
        if like_count is not None:
            self.like_count = like_count

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "platform": self.platform,
            "videoId": self.video_id,
            "channelId": self.channel_id,
            "viewerCount": self.viewer_count,
            "likeCount": self.like_count,
            "messageQueue": [
                {
                    "id": m.id,
                    "content": m.content,
                    "author": m.author,
                    "timestamp": m.timestamp,
                }
                for m in self.message_queue[-10:]  # Only recent messages
            ],
            "superchatQueue": [
                {
                    "id": m.id,
                    "content": m.content,
                    "author": m.author,
                    "amount": m.superchat_amount,
                    "currency": m.superchat_currency,
                }
                for m in self.superchat_queue[-10:]
            ],
            "streamStartedAt": self.stream_started_at,
            "lastMessageAt": self.last_message_at,
            "silenceDuration": self.silence_duration,
        }

File: apps/server-ts/src/state/stream.ts
import type { StreamMessage } from "@aituber-flow/sdk";

export class StreamContext {
  platform: string | null = null;
  videoId: string | null = null;
  channelId: string | null = null;
  viewerCount = 0;
  likeCount = 0;
  messageQueue: StreamMessage[] = [];
  superchatQueue: StreamMessage[] = [];
  streamStartedAt: string | null = null;
  lastMessageAt: string | null = null;
  maxQueueSize = 100;

  get silenceDuration(): number {
    if (!this.lastMessageAt) return 0;
    const last = new Date(this.lastMessageAt).getTime();
    return (Date.now() - last) / 1000;
  }

  addMessage(message: StreamMessage): void {
    this.messageQueue.push(message);
    this.lastMessageAt = message.timestamp;

    if (message.superchatAmount) {
      this.superchatQueue.push(message);
    }

    if (this.messageQueue.length > this.maxQueueSize) {
      this.messageQueue = this.messageQueue.slice(-this.maxQueueSize);
    }
    if (this.superchatQueue.length > this.maxQueueSize) {
      this.superchatQueue = this.superchatQueue.slice(-this.maxQueueSize);
    }
  }

  popMessage(): StreamMessage | null {
    return this.messageQueue.shift() ?? null;
  }

  popSuperchat(): StreamMessage | null {
    return this.superchatQueue.shift() ?? null;
  }

  updateStats(viewerCount?: number, likeCount?: number): void {
    if (viewerCount !== undefined) this.viewerCount = viewerCount;
    if (likeCount !== undefined) this.likeCount = likeCount;
  }

  toDict(): Record<string, unknown> {
    return {
      platform: this.platform,
      videoId: this.videoId,
      channelId: this.channelId,
      viewerCount: this.viewerCount,
      likeCount: this.likeCount,
      messageQueue: this.messageQueue.slice(-10).map((m) => ({
        id: m.id,
        content: m.content,
        author: m.author,
        timestamp: m.timestamp,
      })),
      superchatQueue: this.superchatQueue.slice(-10).map((m) => ({
        id: m.id,
        content: m.content,
        author: m.author,
        amount: m.superchatAmount,
        currency: m.superchatCurrency,
      })),
      streamStartedAt: this.streamStartedAt,
      lastMessageAt: this.lastMessageAt,
      silenceDuration: this.silenceDuration,
    };
  }
}
